File: 7_kyu/Jumping_Number_Special_Numbers_Series_4.clj
(ns kata)

(defn decimal-digits
  "Generate list of all decimal digits as numbers, least digit first."
  [number]
  (cond
    (neg? number) (decimal-digits (- number))
    (< number 10) (list number)
    :else (conj (decimal-digits (quot number 10)) (mod number 10)))
  )


; abs function was added in Clojure 1.11.0
(defn abs
  "Absolute value"
  [x]
  (if (neg? x) (- x) x)
  )


(defn decimal-distance
  "Distance between two decimal digits."
  [a b]
  ; examples: 3-2 is 1, 0-9 is 1.
  (mod (abs (- a b)) 9)
  )


(defn jumping-number?
  "Checks if the number ia a jumping number? (Distance of all adjacent digits is 1)"
  [number]
  (let [digits (decimal-digits number)]
    ; all distance between digits must be 1
    (apply = (cons 1 (map decimal-distance digits (rest digits))))
    )
  )


(defn jumping-number [number]
  (if (jumping-number? number) "Jumping!!" "Not!!")
  )

File: 7_kyu/Jumping_Number_Special_Numbers_Series_4.rs
// Checks if number is a jumping number.
// Jumping means distance between all adjacent decimal digits is 1.
fn jumping_number(n: u64) -> String {
    // Split number into value of last digit and value of leading digits
    let mut last_digit = n % 10;
    let mut leading = n / 10;
    // Process digit by digit
    while leading > 0 {
        let digit = leading % 10;
        if last_digit + 1 != digit && digit + 1 != last_digit {
            return "Not!!".to_string();
        }
        last_digit = digit;
        leading /= 10;
    }
    "Jumping!!".to_string()
}
